witness.md

```diff
@@ -0,0 +1,598 @@
+pattern. 
+
+Example
+
+```
+                             
+Node(n0) Node(n1) BRANCH(mask) |=>
+BranchNode{MAKE_VALUES_ARRAY(mask, n0, n1)}
+                             ^     ^-- ^--- BOUND NODES
+                             |---- BOUND INSTRUCTION PARAM
+          |------ HELPER CALL ------------|
+|----------------- RESULT ------------------|
+
+```
+
+### Bringing it all together
+
+
+So the full syntax is this:
+
+```
+[GUARD <CONDITION> ...] [ NodeType(bound_variable1)... ] INSTRUCTION[(param1 ...)] |=>
+Node(<HELPER_FUNCTION_OR_COMPUTATION>)
+```
+
+`NodeType` is one of the types of nodes to match. Can also be `Node` to match
+any non-nil node.
+
+Substitution rules MUST be non-ambiguous. Even though, there can be multiple
+substitution rules applicable to the whole witness at the same time, there MUST
+be only one rule that is applicable to a certain position in the witness.
+
+So, the minimal substitution rule is the one for the `HASH` instruction that pushes one hash to the stack:
+```
+HASH(hashValue) |=> HashNode{hashValue}
+```
+
+
+## Helper functions
+
+Helper functions are functions that are used in GUARDs or substitution rules.
+
+Helper functions MUST be pure.
+Helper functions MUST have at least one argument.
+Helper functions MAY have variadic parameters: `HELPER_EXAMPLE(arg1, arg2, list...)`.
+Helper functions MAY contain recursion.
+
+## Instructions
+
+### `LEAF key raw_value`
+
+**Substitution rules**
+
+Replaces the instruction with a `ValueNode` wrapped with a `LeafNode`.
+
+```
+LEAF(key, raw_value) |=> LeafNode{key, ValueNode(raw_value)}
+```
+
+### `EXTENSION key`
+
+Wraps a node to the left of the instruction with an `ExtensionNode`.
+
+**Substitution rules**
+
+```
+Node(node) EXTENSION(key) |=> ExtensionNode{key, node}
+```
+
+### `HASH raw_hash`
+
+Replaces the instruction with a `HashNode`.
+
+**Substitution rules**
+
+```
+HASH(hash_value) |=> HashNode{hash_value}
+```
+
+### `CODE raw_code`
+
+Replaces the instruction with a `CodeNode`.
+
+```
+CODE(raw_code) |=> CodeNode{raw_code}
+```
+
+### `ACCOUNT_LEAF key nonce balance has_code has_storage`
+
+Replaces the instruction and, optionally, up to 2 nodes to the left of the
+instructon with a single `AccountNode` wrapped with a `LeafNode`.
+
+**Substitution rules**
+
+```
+GUARD has_code == true
+GUARD has_storage == true
+
+CodeNode(code) Node(storage_hash_node) ACCOUNT_LEAF(key, nonce, balance, has_code, has_storage) |=>
+LeafNode{key, AccountNode{nonce, balance, storage_root, code}}
+
+---
+
+GUARD has_code == true
+GUARD has_storage == true
+
+HashNode(code) Node(storage_hash_node) ACCOUNT_LEAF(key, nonce, balance, has_code, has_storage) |=>
+LeafNode{key, AccountNode{nonce, balance, storage_root, code}}
+
+---
+
+GUARD has_code == false
+GUARD has_storage == true
+
+Node(storage_root) ACCOUNT_LEAF(key, nonce, balance, has_code, has_storage) |=>
+LeafNode{key, AccountNode{nonce, balance, storage_root, nil}}
```

Comment:
Maybe `nil` should be `RLP(nil)` to help with hashing the trie, but this doc seems to omit hashing so maybe it's not relevant

Reply:
yeah, the first version had hashing, but then it just repeated the yellow paper, so we decided to omit that. in the implementation, we keep 2 lists of values: hashes and nodes